mlir/include/mlir/IR/Attributes.h

```diff
@@ -303,7 +300,7 @@ class AttributeInterface
 protected:
   /// Returns the impl interface instance for the given type.
   static typename InterfaceBase::Concept *getInterfaceFor(Attribute attr) {
-#ifndef NDEBUG
+#ifdef EXPENSIVE_CHECKS
```

Comment:
This is critical to check in asserts builds IMO. The promise is quite useless otherwise...

Reply:
I'm open to your opinion on this! On one hand, I agree. I opened this PR simply for discussion (and to evaluate the impact on mojo). Practically no one builds with EXPENSIVE_CHECKS by default. However, compared to asserts in, say, `isa` or `cast`, this one is realllyyyy expensive. What if we had a separate flag that was also on by default but toggleable? The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. It would be nice to specifically toggle these ones and other "heavy" asserts in MLIR off.

Reply:
Our compiler(s) are under development, so turning off asserts in downstream projects might not be as desirable. Having a slow compiler, however, is also affecting the developer experience.

Reply:
> What if we had a separate flag that was also on by default but toggleable?

That seems fine to me.

> The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. 

Which is WAI I believe: they build in assert mode means they want to catch issues in their project, so this will catch issues with their uses of isa/...

Reply:
It's a balance. Subjectively, I think that asserts in milr are already over the line of being too expensive, but it is a fraught topic to bring up. I can't say whether this is one of the big culprits, but expensive checks in isa costing 3x of overall compile time is both troubling and would explain a lot based on my experience.